src/ucp/stream/stream_recv.c

```diff
@@ -303,117 +282,103 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
+ucp_stream_am_data_process(ucp_worker_t *worker, ucp_ep_ext_stream_t *ep,
+                           ucp_stream_am_data_t *am_data, size_t length,
+                           unsigned am_flags)
 {
+    ucp_recv_desc_t  rdesc_tmp;
+    void            *payload;
     ucp_recv_desc_t *rdesc;
+    ucp_request_t   *req;
+    ssize_t          unpacked;
+
+    rdesc_tmp.length         = length;
+    rdesc_tmp.payload_offset = sizeof(*am_data); /* add sizeof(*rdesc) only if
+                                                    am_data wont be handled in
+                                                    place */
+
+    /* First, process expected requests */
+    if (ep->am_data_len == 0) {
+        while (!ucs_queue_is_empty(&ep->reqs)) {
+            req      = ucs_queue_head_elem_non_empty(&ep->reqs, ucp_request_t,
+                                                     recv.queue);
+            payload  = UCS_PTR_BYTE_OFFSET(am_data, rdesc_tmp.payload_offset);
+            unpacked = ucp_stream_rdata_unpack(payload, rdesc_tmp.length, req);
+            if (ucs_unlikely(unpacked < 0)) {
+                ucs_fatal("failed to unpack from am_data %p with offset %u to request %p",
+                          am_data, rdesc_tmp.payload_offset, req);
+            } else if (unpacked == rdesc_tmp.length) {
+                ucp_request_complete_stream_recv(req, ep, UCS_OK);
+                return UCS_OK;
+            }
+            ucp_stream_rdesc_advance(&rdesc_tmp, unpacked);
+            /* This request is full, try next one */
+            ucp_request_complete_stream_recv(req, ep, UCS_OK);
+        }
+    }
 
-    if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+    ucs_assert(rdesc_tmp.length > 0);
+
+    /* Now, enqueue the rest of data */
+    if (ucs_likely(!(am_flags & UCT_CB_PARAM_FLAG_DESC))) {
+        rdesc = (ucp_recv_desc_t*)ucs_mpool_get_inline(&worker->am_mp);
+        ucs_assertv_always(rdesc != NULL,
+                           "ucp recv descriptor is not allocated");
+        rdesc->length         = rdesc_tmp.length;
+        /* reset offset to improve locality */
+        rdesc->payload_offset = sizeof(*rdesc) + sizeof(*am_data);
+        rdesc->flags          = 0;
+        memcpy(ucp_stream_rdesc_payload(rdesc),
+               UCS_PTR_BYTE_OFFSET(am_data, rdesc_tmp.payload_offset),
+               rdesc_tmp.length);
+    } else {
         /* slowpath */
         rdesc        = (ucp_recv_desc_t *)am_data - 1;
-        rdesc->flags = UCP_RECV_DESC_FLAG_UCT_DESC;
-    } else {
-        /* TODO: optimize if there are posted recvs */
-        rdesc = (ucp_recv_desc_t*)ucs_mpool_get_inline(&worker->am_mp);
-        if (ucs_likely(rdesc != NULL)) {
-            rdesc->flags = 0;
-            memcpy((ucp_stream_am_data_t *)(rdesc + 1) + 1, am_data + 1, length);
-        } else {
-            ucs_fatal("ucp recv descriptor is not allocated");
-        }
+        rdesc->length         = rdesc_tmp.length;
```

Comment:
why is rdesc->length set here? it seems like line 343 overrides it
also, maybe just do `rdesc->flags = am_flags` outside the if/else to reduce 1 code line
also could indent the '=' on same column as the line before (rdesc = ... )

Reply:
Most probably rebasing issue... will recheck, thanks!

Reply:
`rdesc->flags = am_flags` will not work  - they have different types

Reply:
right